mytoyota/controller.py

```diff
@@ -72,77 +79,100 @@ async def first_login(self) -> None:
+
+            # Authorise
+            resp = await client.get(
+                self._authorize_endpoint,
+                headers={"cookie": f"iPlanetDirectoryPro={data['tokenId']}"},
             )
-            if response.status_code == HTTPStatus.OK:
-                result: dict[str, Any] = response.json()
-
-                if TOKEN not in result or UUID not in result[CUSTOMERPROFILE]:
-                    raise ToyotaLoginError("Could not get token or UUID from result")
-
-                _LOGGER.debug("Extracting token from result")
-
-                token = result.get(TOKEN)
-
-                if is_valid_token(token):
-                    _LOGGER.debug("Token is the correct format")
-                    self._uuid = result[CUSTOMERPROFILE].get(UUID)
-                    self._token = token
-                    _LOGGER.debug("Saving token and uuid")
-                    self._token_expiration = datetime.now()
-
-            elif response.status_code == HTTPStatus.BAD_GATEWAY:
-                if retry:
-                    await self._update_token(retry=False)
-                    return
-                raise ToyotaApiError("Servers are overloaded, try again later")
-            else:
-                raise ToyotaLoginError(
-                    f"Login failed, check your credentials! {response.text}"
+            if resp.status_code != HTTPStatus.FOUND:
+                _LOGGER.error(
+                    f"Authorization failed:\n"
+                    f"  Status_Code = {resp.status_code}\n"
+                    f"  Text        = {resp.text}\n"
+                    f"  Headers     = {resp.headers}"
                 )
+                raise ToyotaLoginError("Authorization failed")
+            authentication_code = parse.parse_qs(
+                httpx.URL(resp.headers.get("location")).query.decode()
+            )["code"]
+
+            # Retrieve tokens
+            resp = await client.post(
+                self._access_token_endpoint,
+                headers={"authorization": "basic b25lYXBwOm9uZWFwcA=="},
+                # f"basic {self.BASIC_AUTH_STRING}"},
+                data={
+                    "client_id": "oneapp",
+                    "code": authentication_code,
+                    "redirect_uri": "com.toyota.oneapp:/oauth2Callback",
+                    "grant_type": "authorization_code",
+                    "code_verifier": "plain",
+                },
+            )
+            if resp.status_code != HTTPStatus.OK:
+                _LOGGER.debug(
+                    f"Authorization failed:\n"
+                    f"  Status_Code = {resp.status_code}\n"
+                    f"  Text        = {resp.text}\n"
+                    f"  Headers     = {resp.headers}"
+                )
+                raise ToyotaLoginError("Failed to retrieve required tokens")
+
+            access_tokens: dict[str, Any] = resp.json()
+            assert "access_token" in access_tokens and "id_token" in access_tokens
```

Comment:
Better use proper error handling here instead of assertion:

```python
            if "access_token" not in access_tokens or "id_token" not in access_tokens:
                raise ToyotaLoginError("Missing tokens in response")
```